ssl/statem/extensions.c

```diff
@@ -939,9 +939,14 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
      */
     if (s->server) {
         if (!sent) {
-            /* Nothing from the client this handshake; cleanup stale value */
-            OPENSSL_free(s->ext.hostname);
-            s->ext.hostname = NULL;
+            /*
```

Comment:
This branch does nothing, and should continue to do so... Far simpler to just change the outer `if` to:
```
if (sent && s->server && ret == SSL_TLSEXT_ERR_OK && (!s->hit || SSL_IS_TLS13(s)) { ...
```
leaving just the branch that does something in place.

Reply:
I am not fully sold on the assertion that it "should continue to do so [indefinitely]".
I believe that Matt has a similar sense as me, that, in a clean-world view, an application that cares about a supplied-but-not-negotiated SNI value should be responsible for noting that value during the handshake and doing its own storage for that value.  The TLS implementation has no reason to keep around such an ignored value -- how is this different from how we discard the CLIENTHELLO_MSG element linked off the SSL once we're done with processing it?
I agree that it should continue to do so *for now*, since as a core library we have some obligation to avoid gratuitous breaking behavior changes.  But I also think that we need to have some freedom to make breaking changes for sake of library cleanliness and maintainability.

Reply:
I'm going to be blunt and say that you (and perhaps Matt) are simply mistaken. One must need to implement a complex callback song and dance just to have access to the SNI hint sent by the client.  It can be useful for more than just certificate selection. If you want to know whether the SNI was accepted, look in the session, the SSL-layer interface exposes what the client sent.

For example, we have two functions to look a the server certificate chain, one that returns the raw certificate list from the server, and another that returns the validated list. Here, if there's demand we can add a second `SSL_mumble` function that returns the "negotiated" SNI value, by extracting it from the session handle. The present `SSL_get_servername()` function will continue to return what the client sent.

Reply:
> SNI hint

I think this is where the fundamental disagreement lies.  From where I stand, the SNI value sent by the client is not a *hint*, it is an *offer* as part of a *negotiation*.  If the server does not accept the negotiation, the client should not expect any differentiated behavior based on the supplied value, and the server should not change its behavior on the basis of the supplied value without indicating that the negotiation has succeeded.  This is somewhat orthogonal to whether the server software needs to "implement a complex callback song and dance" to do so; we as the library implementation could certainly provide other options for the server to assent to some form of default negotiation, such as if the default certificate would be valid for the client's offer under [conditions TBD].

I am receptive to your arguments about looking in the session for finding out what/whether it was accepted, but I do not think we are going to find a consensus outcome if the above disagreement is not resolved.

Reply:
Not all extensions are alike. SNI conveys information from client to server, which the server may use at its discretion. The server may acknowledge successful use of SNI to perform some non-default action, or it may silently make use of the information, perhaps just for forensic logging. The server's ACK of the SNI extension is largely cosmetic. Some clients may get picky about failure to ACK SNI, and corresponding application servers will implement the callbacks and send ACKs as appropriate. Other applications may take a laissez-faire approach.

Unlike extensions like EtM (Encrypt-then-MAC), SNI has no influence on the rest of the TLS handshake beyond certificate selection. If the client does not care about server ACKs, there's no meaningful "negotiation", and the server is free to use the data as it sees fit.  There's no advantage to pedantic hiding of the SNI data the server may not have acknowledged.

Sure, perhaps some day we can provide a cheap mechanism to ACK all SNI signals while using a default certificate chain and no callbacks, but that should be a pre-requisite to having access to the value sent by the client.

Reply:
To be clear, my claim that it is an offer as part of a negotiation is supported by RFC 8446, which clearly states in Section 4.2:

```
   Extensions are generally structured in a request/response fashion,
   though some extensions are just indications with no corresponding
   response.
```
and follows up with server_name being valid in the CH and EE extension fields (i.e., request and response).  If the server does not supply a response, the extension is not considered negotiated and the corresponding functionality is not to be utilized on the connection in question.

Reply:
We must not let the general theory cloud the specific cases.  While SNI includes a server response to optionally let clients know that the SNI value was used in some fashion (matched exactly? matched to a parent domain? looked at and not matched, and the default chain was returned? ... who knows?) provided the handshake continues and the server certificate chain is acceptable to the client, the SNI ACK/NACK is really of little to no consequence.

Nothing in the protocol precludes the server from using the SNI hostname in whatever way it sees fit, whether or not the value is acknowledged.

For interoperability, servers should ACK matching SNI values when they happen to locate a "matching" chain, that's all.

Reply:
[I'm going to apply a rate limit to my postings here so it does not devolve into the two of us having a heated debate without input from the rest of the team]

Reply:
@kaduk I don't read things the way you have stated them in terms of SNI handling. The server is free to use or ignore the SNI value, the server is free to reject the connection (abort the handshake) or continue if it is ignoring the SNI value.
I don't see any text to indicate what the SNI value in EE is meant to indicate anywhere - is there something I'm missing?

As I read it you are leaping beyond the specification in your interpretation.

Reply:
This sounds like a largely theoretical discussion. We can have the debate about the rights and wrongs of allowing a non-acknowledged SNI value to be available post-handshake at another time. At least for now we need to retain ABI compat so it must be available.

This PR is on the critical path for the 1.1.1 release next week so I don't want it to get derailed. With that in mind @vdukhovni's original comment about refactoring this "if" statement seems reasonable to me.

Reply:
@mattcaswell do you propose that we defer the "largely theoretical" discussion indefinitely, shunt it off to a dedicated issue, or something else?
I will refactor the 'if' statement, but do not have any immediate ideas for why the current change did not seem to resolve the OP's issue (at #7014)

Reply:
I'd suggest opening a separate issue for it